### ISSUE TITLE
MIDI-IN problems since ed7a2c6d4028c307cf0cee653aacd6b99a10265a 

### ISSUE THREAD

**Issue by dcoredump** [moddevices/mod-ttymidi#6]
Hi,

we are using mod-ttymidi for Zynthian. After changing to the last commit I am getting massive drops of MIDI events with my Kawai-MP5 as a master keyboard. This happens only after commit ed7a2c6d4028c307cf0cee653aacd6b99a10265a (Rewrote the serial to MIDI parsing. Now single byte system messages...).

Here are two videos describing the problem:
Before the commit: https://youtu.be/weSIips2qBg
... and after (current master): https://youtu.be/boGhazyUGAw

Not all users of Zynthian are having those problems... it seems to be a combination of the commit AND my master keyboard.

Regards, Holger

**Comment by falkTX** [comment]
Hello there.

Are you able to somehow dump the data this device is sending?
I guess a specific combination of bytes is not being handled correctly by the new code.
Thanks

**Comment by dcoredump** [comment]
Hi,

yep - I will try to catch some events. Perhaps not today (short of time).

Thaks, Holger

**Comment by dcoredump** [comment]
Here are some demos direct recorded from Zynthian:

```
jack_capture_01.mid - recorded with mod-ttymidi (commit ed7a2c6)
jack_capture_02.mid - recorded with mod-ttymidi (HEAD)
jack_capture_03.mid - recorded with mod-ttymidi (HEAD)
```


[jack_captures.zip](https://github.com/moddevices/mod-ttymidi/files/3227885/jack_captures.zip)

Regards, Holger

**Comment by jkbd** [comment]
Hi Holger,
could you please provide some more information for me?
I assume this keyboard is a Kawai MP8. Is there a difference if you connect it as a USB-MIDI device?
Is there a way to make sure the keyboard is not sending SysEx messages?

**Comment by dcoredump** [comment]
Hi Jakob,

My piano is a Kawai MP5 (the little brother of the MP8). In USB mode mod-ttymidi works fine. I just made another MIDI recording - this time with a windows recording software. The interesting part is, that I also recorded the "boot" of the MP5: it sends SYSEX and massive MIDI-CC.

[midi_export.mid.zip](https://github.com/moddevices/mod-ttymidi/files/3228907/midi_export.mid.zip)

So I tried the following: Start MP5, wait until "boot" is finished, than start Zythian (mod-ttymidi): same problems as before...

Until now I Havn't found any information how to disable SYSEX/MIDI-CC at startup...

Regards, Holger

Regards, Holger

**Comment by jkbd** [comment]
Thank you very much!
I confirm there is a problem with the code, because with fuzzy testing I was even able to crash the JACK server. I will be working to solve this tomorrow.

**Comment by jkbd** [comment]
JACK only crashes if the data rate is higher than the MIDI standard requires.

**Comment by dcoredump** [comment]
What does MIDI standard rate mean? 31250 bit/s?

**Comment by jkbd** [comment]
Yes. Which is also 3125 Bytes per second, due to 1 start bit, 8 data bits and 1 stop bit.
Can you please try again with `654164020cdf704872dc470f2108ed11043bb3bf`?

**Comment by dcoredump** [comment]
Hi Jakob,

thanks for patching but still the same problems :-(

Regards, Holger

**Comment by jkbd** [comment]
Sorry, I was just checking with single events and your bootup MIDI recording by ear. I will let you know when I worked on that again. Not before Tuesday.
Best regards
Jakob

**Comment by jkbd** [comment]
I translated your `midi_export.mid` into this
https://github.com/moddevices/mod-ttymidi/blob/master/test/serial-fuzzy-write.py#L139-L176
(To get this byte array, I played back the file with `aplaymidi` over DIN MIDI port, used the `--printonly` mode of TTYMIDI and some text formatting.)

When I test with this on a PC, I find that every event is handled in TTYMIDI and all events, except SysEx, are also sent to JACK (where I monitor with `jack_midi_dump`).

I also tested on ARM (MOD Duo X) what JACK is receiving and the result is the same: Only SysEx is filtered on purpose.

**Comment by dcoredump** [comment]
Ok, that's interesting... the strange thing is: it seems that my Kawai-MP5 is currently the only candidate which does not work with the changes since ed7a2c6. I'm gonna test ma with the quirky no-name master keyboard. Probably it works.

So the big question is: What makes the Kawai-MP5 different?

**Comment by jkbd** [comment]
I don't know what's different about the Kawai. I noticed in your file is a `0xC0 0x00` event -- that is a Control Change with one data byte missing, if I am not mistaken. But it does not cause harm on the long term.
The purpose of the changes since ed7a2c6d4028c307cf0cee653aacd6b99a10265a is to allow MIDI Beat Clock to be handled. Before that, `0xF8` and other single-byte events were filtered.

How to progress?

1. You could try to drop all MIDI messages with no data bytes in this range of code: https://github.com/moddevices/mod-ttymidi/blob/master/src/ttymidi.c#L485-L520. If it works then, it should be possible to define a byte sequence, that triggers the problem.
2. After `0xC0 0x00` the following event gets eaten (which is indeed not perfect for a parser). If there was an invalid event (with multiple bytes) before each NOTE OFF, the behavior would be similar to the one you observed (in the videos). Only the data is not like that.

Let me quickly explain one option to test. Start `socat` like this:
```
$ socat -d -d pty,raw,echo=0 pty,raw,echo=0
2019/05/25 00:42:00 socat[10509] N PTY is /dev/pts/5
2019/05/25 00:42:00 socat[10509] N PTY is /dev/pts/6
2019/05/25 00:42:00 socat[10509] N starting data transfer loop with FDs [5,5] and [7,7]
```
Unfortunately these numbers `5`, `6` will be different. Then you can use TTYMIDI on one end
```
$ ./ttymidi -s /dev/pts/5  # needs JACK running
```
and the Python script to test on the other end:
```
$ python test/serial-fuzzy-write.py  # adjust the `/dev/pts/*` device name in the code
```
Of course, you can use `jack_midi_dump` connected to TTYMIDI to look at the events on the JACK side.

**Comment by falkTX** [comment]
can you verify with latest git if this is fixed now?
MIDI running status was not implemented (I remember doing something about it before, oh well...)
That might have been the cause for this.
Thanks

**Comment by dcoredump** [comment]
I can confirm - it works again with my Kawai MP5! Thanks a lot!!!

**Comment by falkTX** [comment]
Cool, I am closing the issue then.
